File: 2767. Partition String Into Minimum Beautiful Substrings/Solution.py
class Solution:
    def minimumBeautifulSubstrings(self, s: str) -> int:
        self.low = 16
        self.r(s, 0, "", 0)
        return self.low if self.low < 16 else -1
        
    def r(self, s, curr, x, count):
        if curr == len(s):
            if not x:
                self.low = min(self.low, count)
            else:
                if int(x, 2) == 0: return
                if int(x, 2) == 1 or self.check(int(x, 2)):
                    self.low = min(self.low, count + 1)
            return
        if not x and s[curr] == '0': return
        if (not x and s[curr] == '1') or (x and self.check(int(x + s[curr], 2))): self.r(s, curr + 1, "", count + 1)
        self.r(s, curr + 1, x + s[curr], count)
    
    def check(self, x):
        if x == 0: return False
        while x > 1:
            if x % 5 != 0: return False
            x //= 5
        return True
        

File: 2767. Partition String Into Minimum Beautiful Substrings/Solution.cpp
class Solution {
public:
    int low = 16;
    int minimumBeautifulSubstrings(string s) {
        r(s, 0, "", 0);
        return (low < 16) ? low : -1;
    }

    void r(string s, int curr, string x, int count) {
        if (curr == s.length()) {
            if (x == "") {
                low = min(low, count);
            } else {
                if (check(stoi(x, nullptr, 2))) low = min(low, count + 1);
            }
            return;
        }
        if (x == "" && s[curr] == '0') return;
        x += s[curr];
        if (check(stoi(x, nullptr, 2))) r(s, curr + 1, "", count + 1);
        r(s, curr + 1, x, count);
    }

    bool check(int x) {
        if (x == 0) return false;
        while (x > 1) {
            if (x % 5 != 0) return false;
            x /= 5;
        }
        return true;
    }
};
